notebooks/agentic_rag_pipeline_with_memory_and_chat.ipynb

```diff
@@ -0,0 +1,530 @@
+              "🛤️ Connections\n",
+              "  - retriever.documents -> prompt_builder.documents (List[Document])\n",
+              "  - prompt_builder.prompt -> llm.messages (List[ChatMessage])\n",
+              "  - llm.replies -> output_adapter.chat_messages (List[ChatMessage])\n",
+              "  - memory_retriever.documents -> prompt_builder.memories (List[Document])\n",
+              "  - output_adapter.output -> memory_writer.documents (List[Document])"
+            ]
+          },
+          "metadata": {},
+          "execution_count": 26
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "## Visualize the pipeline\n",
+        "\n",
+        "If you want to see a diagram of the pipeline, running this cell will create a file locally. https://docs.haystack.deepset.ai/docs/visualizing-pipelines"
+      ],
+      "metadata": {
+        "id": "zJ49TG4OxZ3J"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "pipeline.show()"
+      ],
+      "metadata": {
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 141
+        },
+        "id": "fgZhV4-4yB1j",
+        "outputId": "012dd8ce-b35a-47c8-c65c-b080d863c2de",
+        "collapsed": true
+      },
+      "execution_count": 1,
+      "outputs": [
+        {
+          "output_type": "error",
+          "ename": "NameError",
+          "evalue": "name 'pipeline' is not defined",
+          "traceback": [
+            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
+            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
+            "\u001b[0;32m<ipython-input-1-49bbe8939f1e>\u001b[0m in \u001b[0;36m<cell line: 1>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mpipeline\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshow\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
+            "\u001b[0;31mNameError\u001b[0m: name 'pipeline' is not defined"
+          ]
+        }
+      ]
+    },
+    {
+      "cell_type": "markdown",
+      "source": [
+        "# Bring it all together\n",
+        "\n",
+        "What follows is glue code to run the pipeline in a loop, provide input instructions, and break when the user enters `Q` to quit.\n",
+        "\n",
+        "Here are some example questions to get you started:\n",
+        "- _How many languages are there?_\n",
+        "- _What is the one with most native speakers_\n",
+        "- _Do you remember the two answers you gave to me before?_"
+      ],
+      "metadata": {
+        "id": "uFwkDCbAk-P8"
+      }
+    },
+    {
+      "cell_type": "code",
+      "source": [
+        "while True:\n",
+        "    question = input(\"Enter your question or Q to exit. Example: How many languages are there?\\n🔮 \")\n",
```

Comment:
I would leave out "Example: How many languages are there?" here because this is otherwise shown repeatedly in every step of the conversation.